cranelift/frontend/src/frontend.rs

```diff
@@ -610,11 +630,186 @@ impl<'a> FunctionBuilder<'a> {
+                        max_vals_in_stack_map_by_size[index] =
+                            core::cmp::max(max_vals_in_stack_map_by_size[index], chunk.len());
+                    }
+
+                    let old_val = safepoints.insert(inst, live);
+                    debug_assert!(old_val.is_none());
+                }
+
+                // (3) Add all needs-stack-map values that are operands to this
+                // instruction to the live set.
+                for val in self.func.dfg.inst_values(inst) {
+                    if self.func_ctx.needs_stack_map.contains(val) {
+                        live.insert(val);
+                    }
+                }
+
+                option_inst = self.func.layout.prev_inst(inst);
+            }
+
+            // After we've processed this block's instructions, remove its
+            // parameters from the live set. This is part of step (1).
+            for val in self.func.dfg.block_params(block) {
+                live.remove(val);
+            }
+        }
+
+        // Create a stack slot for each size of needs-stack-map value. These
+        // slots are arrays capable of holding the maximum number of same-sized
+        // values that must appear in the same stack map at the same time.
+        //
+        // This is indexed by the log2 of the type size.
+        let mut stack_slots = [PackedOption::<ir::StackSlot>::default(); 5];
+        for (log2_size, capacity) in max_vals_in_stack_map_by_size.into_iter().enumerate() {
+            if capacity == 0 {
+                continue;
+            }
+
+            let size = 1usize << log2_size;
+            let slot = self.func.create_sized_stack_slot(ir::StackSlotData::new(
+                ir::StackSlotKind::ExplicitSlot,
+                u32::try_from(size * capacity).unwrap(),
+                u8::try_from(log2_size).unwrap(),
+            ));
+            stack_slots[log2_size] = Some(slot).into();
+        }
+
+        // Insert spills to our new stack slots before each safepoint
+        // instruction.
+        let mut cursor = FuncCursor::new(self.func);
+        for (inst, live_vals) in safepoints {
+            cursor = cursor.at_inst(inst);
+
+            // The offset within each stack slot for the next spill to that
+            // associated stack slot.
+            let mut stack_slot_offsets = [0; 5];
+
+            for val in live_vals {
+                let ty = cursor.func.dfg.value_type(val);
+                let size_of_val = ty.bytes();
+
+                debug_assert!(size_of_val.is_power_of_two());
+                let index = size_of_val.ilog2();
+                let index = usize::try_from(index).unwrap();
+
+                let slot = stack_slots[index].unwrap();
+
+                let offset = stack_slot_offsets[index];
+                stack_slot_offsets[index] += size_of_val;
+
+                cursor
+                    .ins()
+                    .stack_store(val, slot, i32::try_from(offset).unwrap());
```

Comment:
Reloads are conspicuously absent here -- are we making the assumption that we won't have a moving GC (so we need to root but don't need to allow mutation of references)? Would be good to have that in a module-level doc-comment too :-)

Reply:
From the commit message / top level PR comment:

> Here is what is *not* handled yet, and left for future follow up commits:
>
> [...]
>
> * Supporting moving GCs. Right now we generate spills into stack slots for live GC references just before safepoints, but we don't reload the GC references from the stack upon their next use after the safepoint. This involves rewriting uses of the old, spilled values which could be a little finicky, but we think we have a good approach.

Happy to add a comment in the meantime.

Reply:
Definitely missed/forgot that, sorry! Yep, a doc-comment point saying the same would be great.